File: src/main/java/at/searles/parsing/combinators/ParserToReducer.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.printing.PartialConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;

/**
 * Creates a reducer out of a parser.
 */
public class ParserToReducer<T, U, V> implements Reducer<T, V> {

    private final Parser<U> parent;
    private final Fold<T, U, V> fold;

    public ParserToReducer(Parser<U> parent, Fold<T, U, V> fold) {
        this.parent = parent;
        this.fold = fold;
    }

    public V parse(ParserStream stream, @NotNull T left) {
        // must preserve start position.
        long leftStart = stream.start();

        U right = parent.parse(stream);

        if (right == null) {
            return null;
        }

        stream.setStart(leftStart);

        return fold.apply(stream, left, right);
    }

    @Override
    public PartialConcreteSyntaxTree<T> print(@NotNull V v) {
        U right = fold.rightInverse(v);

        if (right == null) {
            return null;
        }

        T left = fold.leftInverse(v);

        if (left == null) {
            return null;
        }

        ConcreteSyntaxTree rightOutput = parent.print(right);

        if (rightOutput == null) {
            return null;
        }

        return new PartialConcreteSyntaxTree<>(left, rightOutput);
    }

    @Override
    public boolean recognize(ParserStream stream) {
        return parent.recognize(stream);
    }

    @Override
    public String toString() {
        return parent.toString() + " >> " + fold.toString();
    }
}

File: src/main/java/at/searles/parsing/combinators/ParserToReducer.kt
package at.searles.parsing.combinators

import at.searles.parsing.Fold
import at.searles.parsing.Parser
import at.searles.parsing.ParserStream
import at.searles.parsing.Reducer
import at.searles.parsing.printing.PartialTree

/**
 * Creates a reducer out of a parser.
 */
class ParserToReducer<T, U, V>(private val parent: Parser<U>, private val fold: Fold<T, U, V>) : Reducer<T, V> {

    override fun parse(left: T, stream: ParserStream): V? {
        val right = stream.parse(parent) ?: return null
        return fold.apply(stream, left, right)
    }

    override fun print(item: V): PartialTree<T>? {
        val right = fold.rightInverse(item) ?: return null
        val left = fold.leftInverse(item) ?: return null
        val rightOutput = parent.print(right) ?: return null
        return PartialTree(left, rightOutput)
    }

    override fun recognize(stream: ParserStream): Boolean {
        return parent.recognize(stream)
    }

    override fun toString(): String {
        return "$parent.fold($fold)"
    }
}